azure/Guidelines.md

```diff
@@ -0,0 +1,861 @@
+
+## Common API Patterns
+
+### Performing an Action
+The REST specification is used to model the state of a resource, and is primarily intended to handle CRUD (Create, Read, Update, Delete) operations. However, many services require the ability to perform an action on a resource, e.g. getting the thumbnail of an image, sending an SMS message.
+
+> :white_check_mark: **DO** pattern your URL like this to perform an action on a resource
+**URL Pattern** 
+ ```https://.../<resource-collection>/<resource-id>/:<action>?<input parameters>```
+
+**SMS Example**
+
+ ```https://.../users/Bob/:send-sms?Text="Hello"```
+
+**Equivalent to (in C#)**
+```users["Bob"].SendSms("Hello")```
+
+> :white_check_mark: **DO** use a POST operation for any action on a resource. 
+>
+> :white_check_mark: **DO** support the Repeatability-Request-ID & Repeatability-First-Sent request headers if the action needs to be idempotent if retries occur.
+
+> :ballot_box_with_check: **YOU SHOULD** use a verb to name your action.
+
+
+### Collections
+> :white_check_mark: **DO** structure the response to a list operation as an object with a top-level array field containing the set (or subset) of resources.
+
+> :ballot_box_with_check: **YOU SHOULD** support paging today if there is ever a chance in the future that the number of items can grow to be very large. 
+> NOTE: It is a breaking change to add paging in the future
+
+> :heavy_check_mark: **YOU MAY** expose an operation that lists your resources by supporting a GET method with a URL to a resource-collection (as opposed to a resource-id).
+
+**Example Response Body** 
+```
+{
+    "value": [
+       { "id": "Item 01", "etag": "0xabc", "price": 99.95, "sizes": null },
+       { … },
+       { … },
+       { "id": "Item 99", "etag": "0xdef", "price": 59.99, "sizes": null }
+    ],
+    "nextLink": "{opaqueUrl}"
+ }
+```
+
+> :white_check_mark: **DO** include the id field and etag field (if supported) for each item as this allows the customer to modify the item in a future operation.
+> 
+> :white_check_mark: **DO** clearly document that resources may be skipped or duplicated across pages of a paginated collection unless the operation has made special provisions to prevent this (like taking a time-expiring snapshot of the collection).
+>
+> :white_check_mark: **DO** return a ```nextLink``` field with a URL that the client can GET in order to retrieve the next page of the collection.
+
+> :ballot_box_with_check: **YOU SHOULD** use ```value``` as the name of the top-level array field unless a more appropriate name is available.
+
+> :no_entry: **DO NOT** return the ```nextLink``` field at all when returning the last page of the collection. 
```

Comment:
Should we make it explit, when we suggest people to support this format even if they are not pageable yet, that it means their only page is implictly "the last page", and therefore it's mostly modeling the API like that (IOW, with "value" but no "nextLink")?

Reply:
@lmazuel -- The way we're interpreting this is... "If you only return one page, it's implicitly your last page." In this case, there would be no "nextLink" field. Is what you are suggesting that services should document their API "as if" it was always "pageable"?

If we got the gist right, would you mind suggesting some text as a starting point for us?

Reply:
Decision: We think this is sufficient at this point.

Reply:
Nope, your interpretation is what I meant, sorry if I wasn't clear :)

Reply:
We know of several listing operations that are not now and never will be pageable. For example, returning a set of countries or states. Because the client code to process a collection (loop for each page) is inconvenient, we do NOT want to force all list operations to support this pattern which would then require that all customer code write loops.